### PR TITLE
test: migrate cli unit/e2e tests to vitest

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Consolidate the testing frameworks and migrate to vitest.

**Description**

Migrate `cli` unit/e2e tests to vitest.

**Steps to test or reproduce**

- Run all tests

### CODE REVIEW

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts [hunk 1]
```diff
@@ -1,7 +1,12 @@
+import path from "node:path";
 import {spawn, Pool, Worker, ModuleThread, QueuedTask} from "@chainsafe/threads";
 import {DecryptKeystoreArgs, DecryptKeystoreWorkerAPI} from "./types.js";
 import {maxPoolSize} from "./poolSize.js";
 
+// Worker constructor consider the path relative to the current working directory
+const workerDir =
+  process.env.NODE_ENV === "test" ? "../../../../../lib/cmds/validator/keymanager/decryptKeystores" : "./";
```
Comment (wemeetagain) [inline packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: why is this necessary?

Comment (nazarhussain) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: Mocha and Vitest handles the runtime dependencies differently. Mocha passes everything through ts-node where it translates internally to right compiled file. While vitest not ties to to load the worker with exact path given. If we don't do that it tries to load `./worker.js` which does not exits in that directory `./worker.ts` exists. 

So why we have to point the worker path to `lib` path in test env.

File: packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts [hunk 2]
```diff
@@ -1,90 +1,94 @@
...
-    it("decrypt keystores if lockfiles already exist if ignoreLockFile=true", async () => {
-      await decryptKeystoreDefinitions(definitions, {logger: console, signal, cacheFilePath});
-      // lockfiles should exist after the first run
+    function testDecryptKeystoreDefinitions(cacheFilePath?: string): void {
+      it("decrypt keystores", async () => {
+        const signers = await decryptKeystoreDefinitions(definitions, {logger: console, signal, cacheFilePath});
+        expect(signers.length).toBe(secretKeys.length);
+        for (const signer of signers) {
+          const hexSecret = signer.secretKey.toHex();
+          // secretKeys doesn't include ${hexSecret}
```
Comment (nflaig) [inline packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts]: I know that this happens due to using an automated tool to migrate the tests but these comments are really unnecessary and don't make sense if the code one line below literally says the same thing.

We should at least take a quick look at those by going through the code and delete the comments that do not add any additional information.

File: packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts [hunk 3]
```diff
@@ -1,90 +1,94 @@
...
 
-      await decryptKeystoreDefinitions(definitions, {logger: console, signal, cacheFilePath, ignoreLockFile: true});
-    });
-  }
-});
+        await decryptKeystoreDefinitions(definitions, {logger: console, signal, cacheFilePath, ignoreLockFile: true});
+      });
+    }
+  },
+  {timeout: 100_000}
```
Comment (nflaig) [inline packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts]: I think we could also just pass the timeout as a number here [see](https://stackoverflow.com/a/68812335/10577550). This would prevent that extra indentation.

or why don't we use `vi.setConfig({testTimeout: 100_000})`?

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts (hunk 1 above)
Comment (nflaig) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: Have you tried if `new URL('./worker.js', import.meta.url)` works?

Comment (nazarhussain) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: Yes tried that it didn't work. `import.meta.url` points to the current `src` directory.

File: packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts [hunk 4]
```diff
@@ -43,11 +41,9 @@ describe("decryptKeystoreDefinitions", () => {
     const cacheFilePath = path.join(dataDir, "cache", "keystores.cache");
 
     beforeEach(async () => {
-      // create cache file to ensure keystores are loaded from cache during tests
```
Comment (nflaig) [inline packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts:46]: Not sure if it is good to remove those comments, it still adds some context.

The comments I meant are the ones created by the vitest migration tool that _literally_ say the same as the code itself.

For example 👇 

```
// secretKeys doesn't include ${hexSecret}
expect(secretKeys.includes(hexSecret)).toBe(true);
```

I think we should not remove any other comments in this refactoring that are not a result of migrating to vitest

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts (hunk 1 above)
Comment (nflaig) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: Can we make this a bit more robust at least right now it breaks if
- someone runs tests without NODE_ENV set (we should avoid on relying on this env to be set) <-- vitest might set this haven't tested
- the file is moved or folders are renamed (this will not be detected in a refactoring that moves files / folders)

File: packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts (hunk 4 above)
Comment (nazarhussain) [reply packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts:46]: I remove the comments from one file only where it's clear from the test case and expectations. All other comments are there for context.

Comment (nflaig) [reply packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts:46]: I commented specifically on this line because it is not that clear here in my opinion

we are calling with the sole purpose to create the cache files beforehand but someone not being familiar with that function might not know that it has the side effect of creating cache files
```ts
 await decryptKeystoreDefinitions(definitions, {logger: console, cacheFilePath, signal});
```

This is quite subjective I guess, but this discussion shouldn't be part of this PR, let's rather do a separate PR to clean up unnecessary comments.

In this PR the only thing that should be cleaned up in my opinion are comments created by the vitest auto migration tooling.

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts (hunk 1 above)
Comment (nflaig) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: @nazarhussain I noticed we are using this pattern already on unstable branch. This looks like a bug in vitest as this workaround is not reasonable in my opinion, as testing framework should not force you to use weird hacks in your src code.

I think what we should do is
- open a issue on vitest repo to get this resolved (if it doesn't exist already)
- create an issue on lodestar repository, we also should make sure to reference all lines that have this in the issue
https://github.com/ChainSafe/lodestar/blob/d3005bf58efb527d2887a922689d74e32358bf79/packages/beacon-node/src/chain/bls/multithread/index.ts#L33
https://github.com/ChainSafe/lodestar/blob/d3005bf58efb527d2887a922689d74e32358bf79/packages/beacon-node/src/network/core/networkCoreWorkerHandler.ts#L33

File: packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts (hunk 4 above)
Comment (nazarhussain) [reply packages/cli/test/unit/validator/decryptKeystoreDefinitions.test.ts:46]: No comments are removed automatically by any tool. For such assertions with extra messages, it ask for the user feedback.

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: I had created an issue for it. https://github.com/ChainSafe/lodestar/issues/6208

It's not a bug in vitest neither a hack. Mocha uses runtime `NODE_OPTIONS` to specify a node loader to `ts-node`, so every code that you run in Node had to go through `ts-node`, that's why it's slow. 

Vitest compile and deliver the code to test runner, on **runtime**, so why when it requests for a file `src/worker.js` it does not find it there.  

Will try to find a neat workaround to address this thing in separate PR. 

> @nazarhussain I noticed we are using this pattern already on unstable branch. This looks like a bug in vitest as this workaround is not reasonable in my opinion, as testing framework should not force you to use weird hacks in your src code.
> 
> I think what we should do is
> 
> * open a issue on vitest repo to get this resolved (if it doesn't exist already)
> * create an issue on lodestar repository, we also should make sure to reference all lines that have this in the issue
>   https://github.com/ChainSafe/lodestar/blob/d3005bf58efb527d2887a922689d74e32358bf79/packages/beacon-node/src/chain/bls/multithread/index.ts#L33
>   
>   https://github.com/ChainSafe/lodestar/blob/d3005bf58efb527d2887a922689d74e32358bf79/packages/beacon-node/src/network/core/networkCoreWorkerHandler.ts#L33

Comment (nflaig) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: > It's not a bug in vitest neither a hack.

Having to use `process.env.NODE_ENV === "test"` in your source code is definitely a hack and `../../../../lib/` as well, if there is no other way to make it work we have to bite the bullet but this is not pretty.

**Review by nflaig — APPROVED** [review]

File: packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: The use of `process.env.NODE_ENV` to `production` or `test` is common practice in JS projects to optimize some logic for particular use cases. And mostly bundling the code such usages are stripped down. 

That `path` looks ugly, because we are not using any import aliases.

Comment (nflaig) [reply packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts:8]: > The use of process.env.NODE_ENV to production or test is common practice in JS projects to optimize some logic for particular use cases

See [NODE ENV=production is a lie](https://www.youtube.com/watch?v=HMM7GJC5E2o), I don't think it good to rely on that. For testing it might be still ok, I would still consider it a hack though.

> That path looks ugly, because we are not using any import aliases.

My main concern is not that it is ugly but rather that it is really fragile and will break if any folder changes in that path.

File: packages/cli/test/utils/runUtils.ts [hunk 5]
```diff
@@ -17,12 +17,20 @@ export function findApiToken(dirpath: string): string {
...
+  }
 }
 
 /**
  * Similar to `expectDeepEquals` but only checks presence of all elements in array, irrespective of their order.
  */
 export function expectDeepEqualsUnordered<T>(a: T[], b: T[], message: string): void {
-  expect(a).to.have.deep.members(b, message);
+  try {
+    expect(a.sort()).toEqual(b.sort());
```
Comment (nflaig) [inline packages/cli/test/utils/runUtils.ts:32]: @nazarhussain  I don't think this works as expected, see [failed test run](https://github.com/ChainSafe/lodestar/actions/runs/7261446669/job/19782733198#step:6:4638).

Note that tests might pass sometimes, there was a previous issue related to that
- https://github.com/ChainSafe/lodestar/pull/5641

Also noticing from that issue, our test output is not really detailed compared to previous output if tests fail (see screenshot in linked issue). Previously the output showed exactly what value was wrong while now it does tell me anything about what part of the data is wrong

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.14.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.14.0) :tada:
